src/Faithlife.OAuth/OAuthUtility.cs

```diff
@@ -124,48 +124,99 @@ public static Uri CreateHmacSha1Uri(Uri uri, string httpMethod, string consumerT
 
-		/// <summary>
-		/// Encode string per OAuth spec (see http://oauth.net/core/1.0/ section 5.1).
-		/// </summary>
-		internal static string PercentEncode(string? value)
+		private static string CreateHmacSha1Signature(Uri uri, string httpMethod, string consumerSecret, string? tokenSecret, ICollection<KeyValuePair<string, string>> parameters) =>
+			CreateHmacSha1Signature(CreateSignatureBase(uri, httpMethod, parameters, out _), consumerSecret, tokenSecret);
+
+		internal static string CreateHmacSha1Signature(string signatureBase, string consumerSecret, string? tokenSecret)
 		{
-			if (value is null)
-				return "";
+			using var hmacsha1 = new HMACSHA1(Encoding.UTF8.GetBytes(CreatePlainTextSignature(consumerSecret, tokenSecret)));
+			return Convert.ToBase64String(hmacsha1.ComputeHash(Encoding.UTF8.GetBytes(signatureBase)));
+		}
 
-			var valueBytes = Encoding.UTF8.GetBytes(value);
+		private static string CreateHmacSha1AuthorizationHeaderValue(Uri uri, string httpMethod, string consumerSecret, string? tokenSecret, ICollection<KeyValuePair<string, string>> parameters) =>
+			CreateHmacShaAuthorizationHeaderValue(uri, httpMethod, consumerSecret, tokenSecret, parameters, CreateHmacSha1Signature(CreateSignatureBase(uri, httpMethod, parameters, out var newUIUri), consumerSecret, tokenSecret));
 
-			var builder = new StringBuilder();
-			foreach (var valueByte in valueBytes)
-			{
-				if (IsUnreserved(valueByte))
-					builder.Append((char) valueByte);
-				else
-					builder.Append('%' + "{0:X2}".FormatInvariant((int) valueByte));
-			}
+		private static string CreateHmacShaAuthorizationHeaderValue(Uri uri, string httpMethod, string consumerSecret, string? tokenSecret, ICollection<KeyValuePair<string, string>> parameters, string signatureMethod)
+		{
+			string signatureBase = CreateSignatureBase(uri, httpMethod, parameters, out var newUri);
 
-			return builder.ToString();
+			parameters.Add(new KeyValuePair<string, string>(OAuthConstants.Signature, signatureMethod));
+
+			return OAuthConstants.HeaderPrefix + " " + parameters
+				.Where(parameter => parameter.Key.StartsWithOrdinal(OAuthConstants.ParameterPrefix) && parameter.Value != null)
+				.Select(p => "{0}=\"{1}\"".FormatInvariant(p.Key, PercentEncode(p.Value)))
+				.Join(",");
 		}
 
-		internal static string CreateHmacSha1Signature(string signatureBase, string consumerSecret, string? tokenSecret)
+		/// <summary>
+		/// Create OAuth authentication header using the HMACSHA256 signature method.
+		/// </summary>
+		public static string CreateHmacSha256AuthorizationHeaderValue(Uri uri, string httpMethod, string consumerToken, string consumerSecret, string? accessToken = null, string? accessSecret = null, string? callback = null, string? verifier = null, INonceCreator? nonceCreator = null, ISystemTime? systemTime = null)
```

Comment:
I don't like this being just one `public` method with many optional parameters, because it's not clear at all which parameters go together.

There are three valid cases:

* accessToken & accessSecret
* callback
* temporaryToken, temporarySecret, & verifier

Having three separate overloads helps clarify for the user that these are three separate use cases. Sure, the Intellisense tooltip is a little dense, but it's obvious that if you have a temporaryToken, you also need a temporary secret and verifier. If they're all just optional, this is obscured.

![image](https://user-images.githubusercontent.com/188129/127898105-bab2d881-f701-4bd7-bcb7-cc521436fc35.png)

Since we only need the `consumerToken`, `consumerSecret`, `accessToken`, `accessSecret` overload, I'm in favour of streamlining and reducing to that use case and not even writing the other overloads.

(The others are only needed if you're doing the "OAuth 1.0a dance" with a HMAC-SHA256 server; we don't have that need and maybe never will.)

Reply:
![image](https://user-images.githubusercontent.com/188129/127898353-3944a5d6-0d7c-4413-bb94-8f8d8b87b6a1.png)

By contrast, it's not obvious what needs to be passed here. And by creating 2+ overloads, we can remove the unhelpful nullability annotations. It's not true that all of these parameters are independently nullable.

```suggestion
		public static string CreateHmacSha256AuthorizationHeaderValue(Uri uri, string httpMethod, string consumerToken, string consumerSecret, string accessToken, string accessSecret, INonceCreator? nonceCreator = null, ISystemTime? systemTime = null)
```